README.md
# SODALITE k8s collection
<!-- Add CI and code coverage badges here. Samples included below. -->
[![CI](https://github.com/mihaTrajbaric/k8s/actions/workflows/ansible-test.yml/badge.svg?event=push)](https://github.com/mihaTrajbaric/k8s/actions/workflows/ansible-test.yml) [![codecov](https://codecov.io/gh/mihaTrajbaric/k8s/branch/main/graph/badge.svg?token=MHBEH17281)](https://codecov.io/gh/mihaTrajbaric/k8s)

<!-- Describe the collection and why a user would want to use it. What does the collection do? -->

This repository hosts the `sodalite.k8s` Ansible Collection.

The collection includes quality Ansible content to help automate the management of applications in Kubernetes

<!--start requires_ansible-->
## Ansible version compatibility

This collection has been tested against following Ansible versions: **>=2.9.17**.

Plugins and modules within a collection may be tested with only specific Ansible versions.
A collection may contain metadata that identifies these versions.
PEP440 is the schema used to describe the versions of Ansible.
<!--end requires_ansible-->

## Python Support

* Collection supports 3.6+

Note: Python2 is deprecated from [1st January 2020](https://www.python.org/doc/sunset-python-2/). Please switch to Python3.

## External requirements

Collection requires python package [kubernetes>=12.0.0](https://pypi.org/project/kubernetes/).

## Included content

<!--start collection content-->
### Modules
Name | Description
--- | ---
[sodalite.k8s.config_map](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.config_map_module.rst)|Creates k8s ConfigMap
[sodalite.k8s.deployment](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.deployment_module.rst)|Creates k8s Deployment
[sodalite.k8s.ingress](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.ingress_module.rst)|Creates k8s Ingress
[sodalite.k8s.namespace](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.namespace_module.rst)|Creates k8s Namespace
[sodalite.k8s.pvc](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.pvc_module.rst)|Creates k8s PersistentVolumeClaim
[sodalite.k8s.secret](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.secret_module.rst)|Creates k8s Secret
[sodalite.k8s.service](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.service_module.rst)|Creates k8s Service
[sodalite.k8s.storage_class](https://github.com/mihaTrajbaric/k8s/blob/main/docs/sodalite.k8s.storage_class_module.rst)|Creates k8s StorageClass

<!--end collection content-->

## Installation and Usage

### Installing the Collection from GIT repository

Before using the Sodalite.8s collection, you need to install it with the Ansible Galaxy CLI:

    ansible-galaxy collection install git+https://github.com/mihaTrajbaric/k8s.git

You can also include it in a `requirements.yml` file and install it via `ansible-galaxy collection install -r requirements.yml`, using the format:

```yaml
---
collections:
  - name: https://github.com/mihaTrajbaric/k8s.git
    type: git
```


### Installing the Collection from Ansible Galaxy (TODO - no published yet)

Before using the Sodalite.8s collection, you need to install it with the Ansible Galaxy CLI:

    ansible-galaxy collection install sodalite.k8s

You can also include it in a `requirements.yml` file and install it via `ansible-galaxy collection install -r requirements.yml`, using the format:

```yaml
---
collections:
  - name: sodalite.k8s
    version: 1.0.0
```

### Installing the Kubernetes Python Library

Content in this collection requires the [Kubernetes Python client](https://pypi.org/project/kubernetes/) to interact with Kubernetes' APIs. You can install it with:

    pip3 install kubernetes>=12.0.0

### Using modules from the Kubernetes Collection in your playbooks

It's preferable to use content in this collection using their Fully Qualified Collection Namespace (FQCN), for example `sodalite.k8s.deployment`:

```yaml
---
- hosts: localhost
  gather_facts: false
  connection: local

  tasks:
    - name: Ensure namespace exists.
      sodalite.k8s.namespace:
        name: test-namespace
        state: present

    - name: Create new configmap
      sodalite.k8s.config_map:
        name: myapp-config
        namespace: test-namespace
        data:
          DB_IP: another-service

    - name: Ensure the myapp Deployment exists
      sodalite.k8s.deployment:
        name: myapp-deployment
        namespace: test-namespace
        state: present
        labels:
          app: myapp
        selector:
          match_labels:
            app: myapp
        containers:
          - name: myapp-container
            image: myapp:latest
            ports:
              - host_port: 80
                container_port: 80
                protocol: TCP
            env_from:
              - config_map:
                  name: myapp-config

    - name: Ensure the myapp Service exists
      sodalite.k8s.service:
        name: myapp-service
        namespace: test-namespace
        state: present
        type: LoadBalancer
        selector:
          app: myapp
        ports:
          - name: my-port
            port: 8080
            target_port: 8080
```

See [Ansible Using collections](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html) for more details.

## Contributing to this collection (TODO)

<!--Describe how the community can contribute to your collection. At a minimum, include how and where users can create issues to report problems or request features for this collection.  List contribution requirements, including preferred workflows and necessary testing, so you can benefit from community PRs. If you are following general Ansible contributor guidelines, you can link to - [Ansible Community Guide](https://docs.ansible.com/ansible/latest/community/index.html). -->


## Release notes

See the [changelog](https://github.com/mihaTrajbaric/k8s/tree/main/CHANGELOG.rst).

<!-- ## Roadmap -->

<!-- Optional. Include the roadmap for this collection, and the proposed release/versioning strategy so users can anticipate the upgrade/update cycle. -->

## Licensing

<!-- Include the appropriate license information here and a pointer to the full licensing details. If the collection contains modules migrated from the ansible/ansible repo, you must use the same license that existed in the ansible/ansible repo. See the GNU license example below. -->

GNU General Public License v3.0 or later.

See [LICENSE](https://www.gnu.org/licenses/gpl-3.0.txt) to see the full text.



CHANGELOG.rst
==========================
Sodalite K8s Release Notes
==========================

.. contents:: Topics


v1.0.0
======

New Modules
-----------

- config_map - Creates k8s ConfigMap
- deployment - Creates k8s Deployment
- ingress - Creates k8s Ingress
- namespace - Creates k8s Namespace
- pvc - Creates k8s PersistentVolumeClaim
- secret - Creates k8s Secret
- service - Creates k8s Service
- storage_class - Creates k8s StorageClass


plugins/modules/ingress.py
#!/usr/bin/python

# Copyright: (c) 2021, Mihael Trajbarič <mihael.trajbaric@xlab.si>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import (absolute_import, division, print_function)

__metaclass__ = type
DOCUMENTATION = r'''
---
module: ingress

short_description: Creates k8s Ingress

version_added: "1.0.0"

description: Creates k8s Ingress, which is s a collection of rules that allow inbound connections to reach the
             endpoints defined by a backend. An Ingress can be configured to give services externally-reachable
             urls, load balance traffic, terminate SSL, offer name based virtual hosting etc.

extends_documentation_fragment:
    - sodalite.k8s.common_update_options
    - sodalite.k8s.metadata_options
    - kubernetes.core.k8s_auth_options
    - kubernetes.core.k8s_wait_options
    - kubernetes.core.k8s_delete_options

options:
    ingress_class_name:
        description:
        - The name of the IngressClass cluster resource.
        - The associated IngressClass defines which controller will implement the resource.
        - This replaces the deprecated kubernetes.io/ingress.class annotation.
         - For backwards compatibility, when that annotation is set, it must be given precedence over this field.
         - The controller may emit a warning if the field and annotation have different values.
        type: str
    default_backend_service:
        description:
        - The service backend that should handle requests that don't match any rule.
        - If I(rules) are not specified, I(default_backend_service) must be specified.
        - If I(default_backend_service) is not set, the handling of requests that do not match any of the rules will be
          up to the Ingress controller.
        type: dict
        suboptions:
            name:
                description:
                - Name of the referenced service.
                - The service must exist in the same namespace as the Ingress object.
                type: str
                required: true
            port:
                description:
                - A port name or port number of the port on the Service.
                - Number must be a valid port number (0 < x < 65536).
                - Name must be an IANA_SVC_NAME.
                type: str
                required: true
    rules:
        description:
        - A list of host rules used to configure the Ingress.
        - If unspecified, or no rule matches, all traffic is sent to the default backend.
        type: list
        elements: dict
        suboptions:
            host:
                description:
                - Host is the fully qualified domain name of a network host, as defined by RFC 3986.
                - IPs and ports are not allowed.
                - Implicit port is set to C(80) (http) and C(443) (https).
                - If I(host) is unspecified, the Ingress routes all traffic based on paths.
                - Host can be "precise" which is a domain name without the terminating dot of a network host (e.g.
                  "foo.bar.com") or "wildcard", which is a domain name prefixed with a single wildcard label (e.g.
                  "*.foo.com").
                - You cannot have a wildcard label by itself (e.g. Host == "*").
                - Some k8s distros do not support wildcard
                type: str
            paths:
                description:
                - A collection of paths that map requests to backends.
                - Incoming urls matching the path are forwarded to the backend.
                required: true
                type: list
                elements: dict
                suboptions:
                    backend_service:
                        description:
                        - The referenced service endpoint to which the traffic will be forwarded to.
                        type: dict
                        required: true
                        suboptions:
                            name:
                                description:
                                - Name of the referenced service.
                                - The service must exist in the same namespace as the Ingress object.
                                type: str
                                required: true
                            port:
                                description:
                                - A port name or port number of the port on the Service.
                                - Number must be a valid port number (0 < x < 65536).
                                - Name must be an IANA_SVC_NAME.
                                required: true
                                type: str
                    path:
                        description:
                        - Path to be matched  against the path of an incoming request.
                        - It should follow "path" part of a URL as defined by RFC 3986, but this is currently
                          not validated.
                        - Must begin with a '/'.
                        - When unspecified, all paths from incoming requests are matched.
                        default: '/'
                        type: str
                    path_type:
                        description:
                        - Determines the interpretation of the I(path) matching.
                        - I(path_type=Exact) matches the URL path exactly.
                        - I(path_type=Prefix) matches based on a URL path prefix split by '/'.
                        type: str
                        default: Prefix
                        choices: [ Exact, Prefix ]
    tls:
        description:
        - TLS configuration.
        - Currently the Ingress only supports a single TLS port, 443.
        - If multiple members of this list specify different hosts, they will be multiplexed on the same port according
          to the hostname specified through the SNI TLS extension, if the ingress controller fulfilling the ingress
          supports SNI.
        type: list
        elements: dict
        suboptions:
            hosts:
                description:
                - A list of hosts included in the TLS certificate.
                - The values in this list must match the name/s used in the tlsSecret.
                - Defaults to the wildcard host setting for the loadbalancer controller fulfilling this Ingress,
                  if left unspecified.
                - TLS will not work on the default rule because the certificates would have to be issued for all the
                  possible sub-domains. Therefore, hosts in the tls section need to explicitly match the host in
                  the rules section.
                type: list
                elements: str
            secret:
                description:
                - The name of the secret used to terminate TLS traffic on port 443.
                - Field is left optional to allow TLS routing based on SNI hostname alone.
                - If the SNI host in a listener conflicts with the "Host" header field used by an IngressRule,
                  the SNI host is used for termination and value of the Host header is used for routing.
                type: str

seealso:
- name: K8s Ingress documentation
  description: Documentation about K8s Ingress resource on kubernetes website
  link: https://kubernetes.io/docs/concepts/services-networking/ingress/
- name: K8s Ingress API reference
  description: API reference for K8s Ingress resource on kubernetes website
  link: https://kubernetes.io/docs/reference/kubernetes-api/service-resources/ingress-v1/

author:
    - Mihael Trajbarič (@mihaTrajbaric)
'''

EXAMPLES = r'''
- name: Create simple ingress with default backend
  sodalite.k8s.ingress:
    name: ingress-minimal
    state: present
    default_backend_service:
      name: default-service
      port: 8080

- name: Ingress with prefix path
  sodalite.k8s.ingress:
    name: ingress-path
    state: present
    rules:
    - host: foo.bar.com
      paths:
        - path: /testpath
          path_type: Prefix
          backend_service:
             name: test
             port: 80

- name: Ingress with TLS configuration
  sodalite.k8s.ingress:
    name: ingress-tls
    state: present
    rules:
    - host: https-foo.bar.com
      paths:
        - path: /testpath
          path_type: Prefix
          backend_service:
             name: service1
             port: app-port
    tls:
      - hosts: ['https-foo.bar.com']
        secret: secret-tls

- name: State absent
  sodalite.k8s.ingress:
    name: ingress-test
    state: absent
'''

RETURN = r'''
result:
  description:
  - The created, patched, or otherwise present object. Will be empty in the case of a deletion.
  returned: success
  type: complex
  contains:
     api_version:
       description: The versioned schema of this representation of an object.
       returned: success
       type: str
     kind:
       description: Represents the REST resource this object represents.
       returned: success
       type: str
     metadata:
       description: Standard object metadata. Includes name, namespace, annotations, labels, etc.
       returned: success
       type: dict
     spec:
       description: Specific attributes of the object.
       returned: success
       type: dict
     status:
       description: Current status details for the object.
       returned: success
       type: dict
     duration:
       description: elapsed time of task in seconds
       returned: when C(wait) is true
       type: int
       sample: 48
     error:
       description: error while trying to create/delete the object.
       returned: error
       type: dict
'''

from ansible_collections.sodalite.k8s.plugins.module_utils.ansiblemodule import AnsibleModule
from ansible_collections.sodalite.k8s.plugins.module_utils.args_common import (update_arg_spec,
                                                                               UPDATE_MUTUALLY_EXCLUSIVE)
from ansible_collections.sodalite.k8s.plugins.module_utils.common import Validators, CommonValidation, Marshalling
from ansible_collections.sodalite.k8s.plugins.module_utils.helper import clean_dict


def definition(params):

    def get_port(port, _type):
        """
        makes sure port gets assigned to 'name' or 'number', depending on IntOrString unmarshalling
        """
        unmarshalled_port = Marshalling.unmarshall_int_or_string(port)
        if isinstance(unmarshalled_port, _type):
            return unmarshalled_port
        return None

    def ingress_backend(backend_service):
        if backend_service is None:
            return None
        return {
            "service": {
                "name": backend_service.get('name'),
                "port": {
                    # function get_port(port, _type) will make sure only one of (name, number) != None
                    # (required by mutually exclusive condition)
                    'name': get_port(backend_service.get('port'), str),
                    'number': get_port(backend_service.get('port'), int),
                }
            }
        }

    body = {
        "apiVersion": "networking.k8s.io/v1",
        "kind": "Ingress",
        "metadata": {
            "name": params.get('name'),
            "labels": params.get('labels'),
            "annotations": params.get('annotations')
        },
        "spec": {
            "defaultBackend": ingress_backend(params.get('default_backend_service')),
            "ingressClassName": params.get('ingress_class_name'),
            "rules": [
                {
                    'host': rule.get('host'),
                    'http': {
                        'paths': [
                            {
                                'backend': ingress_backend(path.get('backend_service')),
                                'path': path.get('path'),
                                'pathType': path.get('path_type')
                            }
                            for path in rule.get('paths') or list()
                        ]
                    }
                }
                for rule in params.get('rules') or list()
            ],
            'tls': [
                {
                    'hosts': tls_config.get('hosts'),
                    'secretName': tls_config.get('secret')
                }
                for tls_config in params.get('tls') or list()
            ]
        }
    }

    return clean_dict(body)


def validate(module, k8s_definition):
    CommonValidation.metadata(module, k8s_definition)

    if not Validators.dns_subdomain(k8s_definition['metadata']['name']):
        module.fail_json(msg=f"'name' {Validators.dns_subdomain_msg}")

    spec = k8s_definition.get('spec', dict())
    if not (spec.get('defaultBackend') or spec.get('rules')):
        module.fail_json(msg="At least one of (default_backend_service, rules) must be present. If present, 'rules'"
                             "must contain at least one element")

    default_backend_service = spec.get('defaultBackend', dict()).get('service', dict())
    if not Validators.dns_label_1035(default_backend_service.get('name')):
        module.fail_json(msg=f"default_backend_service.name {Validators.dns_label_1035_msg}")
    if not Validators.port(default_backend_service.get('port', dict()).get('number')):
        module.fail_json(msg=f"default_backend_service.port can be a port name or number. If it is a port number, "
                             f"it {Validators.port_msg}")
    if not Validators.iana_svc_name(default_backend_service.get('port', dict()).get('name')):
        module.fail_json(msg=f"default_backend_service.port can be a port name or number. If it is a port name, "
                             f"it {Validators.iana_svc_name_msg}")

    for i, rule in enumerate(spec.get('rules', list())):
        if not Validators.dns_subdomain_wildcard(rule.get('host')):
            module.fail_json(msg=f'rules[{i}].host {Validators.dns_subdomain_wildcard_msg}')

        paths = rule.get('http', dict()).get('paths')
        if not paths:
            module.fail_json(msg=f'rules[{i}].paths must contain at least one parameter')

        for j, path in enumerate(paths):
            backend_service = path.get('backend').get('service')
            if not Validators.dns_label_1035(backend_service.get('name')):
                module.fail_json(msg=f"rules[{i}].paths[{j}].backend_service.name {Validators.dns_label_1035_msg}")
            if not Validators.port(backend_service.get('port').get('number')):
                module.fail_json(
                    msg=f"rules[{i}].paths[{j}].backend_service.port can be a port name or number. If it is a port "
                        f"number, it {Validators.port_msg}")
            if not Validators.iana_svc_name(backend_service.get('port').get('name')):
                module.fail_json(
                    msg=f"rules[{i}].paths[{j}].backend_service.port can be a port name or number. If it is a port "
                        f"name, it {Validators.iana_svc_name_msg}")
            if not Validators.url_path(path.get('path')):
                module.fail_json(msg=f"rules[{i}].paths[{j}].path {Validators.url_path_msg}")

    for i, tls_conf in enumerate(spec.get('tls', list())):
        if not Validators.dns_subdomain(tls_conf.get('secretName')):
            module.fail_json(msg=f"tls[{i}].secret {Validators.dns_subdomain_msg}")


def main():
    argspec = update_arg_spec()
    argspec.update(dict(
        ingress_class_name=dict(type='str'),
        default_backend_service=dict(type='dict', options=dict(
            name=dict(type='str', required=True),
            port=dict(type='str', required=True),
        )),
        rules=dict(type='list', elements='dict', options=dict(
            host=dict(type='str'),
            paths=dict(type='list', required=True, elements='dict', options=dict(
                backend_service=dict(type='dict', required=True, options=dict(
                    name=dict(type='str', required=True),
                    port=dict(type='str', required=True),
                )),
                path=dict(type='str', default='/'),
                path_type=dict(type='str', default='Prefix', choices=['Exact', 'Prefix']),
            ))
        )),
        tls=dict(type='list', elements='dict', options=dict(
            hosts=dict(type='list', elements='str'),
            secret=dict(type='str', no_log=False)
        ))
    ))

    module = AnsibleModule(argument_spec=argspec,
                           mutually_exclusive=UPDATE_MUTUALLY_EXCLUSIVE,
                           supports_check_mode=True)
    from ansible_collections.sodalite.k8s.plugins.module_utils.k8s_connector import execute_module

    k8s_def = definition(module.params)
    if module.params.get('state') != 'absent':
        validate(module, k8s_def)

    execute_module(module, k8s_def)


if __name__ == '__main__':
    main()
